### DISCUSSION TITLE
Diversion header missing

### DISCUSSION THREAD

**Discussion by JyrgCodes** [hnimminh/libresbc#181] [Q&A]
Hi @hnimminh 
I have successfully set up the connection Twilio<->LibreSBC<->FreePBX. The calls go through without a problem and most headers are also correct.
When redirecting calls on the FreePBX side, a diversion header is added to the INVITE request. But it seems that this isn't getting through to Twilio. Do I have to add some kind of header manipulation or where could the header be stuck?

The request which leaves FreePBX:
```
INVITE sip:{c-number}@{LibreSBC Private IP}:5080;user=phone SIP/2.0
Via: SIP/2.0/UDP {FreePBX Private IP}:5060;rport;branch=z9hG4bKPj78e21953-41ba-450e-a39f-f7ba2612d2a2
From: "{a-number}" <sip:{a-number}@{FreePBX Private IP};user=phone>;tag=be61ef0d-29bc-42d7-85b6-2a7e5dc3a1a7
To: <sip:{c-number}@{LibreSBC Private IP};user=phone>
Contact: <sip:asterisk@{FreePBX Private IP}:5060>
Call-ID: 85bfe52e-0cff-41ea-9240-ae7f64f163a9
CSeq: 20093 INVITE
Allow: OPTIONS, INVITE, ACK, BYE, CANCEL, UPDATE, PRACK, REGISTER, SUBSCRIBE, NOTIFY, PUBLISH, MESSAGE, REFER
Supported: 100rel, timer, replaces, norefersub, histinfo
Session-Expires: 1800
Min-SE: 90
P-Asserted-Identity: "{a-number}" <sip:{a-number}@{FreePBX Private IP};user=phone>
Diversion: <sip:{b-number}@{FreePBX Private IP}:5060>;reason=unconditional;screen=no;privacy=off
Max-Forwards: 70
User-Agent: FPBX-16.0.40.11(18.16.0)
Content-Type: application/sdp
Content-Length:   229
```

The request which arrives on Twilio:
```
Via: SIP/2.0/UDP {LibreSBC Public IP};rport;branch=z9hG4bKStSc8eSSX6cFm
Max-Forwards: 69
From: "{a-number}" <sip:{a-number}@{LibreSBC Public IP}>;tag=KyQr65KZ30D1K
To: <sip:{c-number}@whispy.pstn.twilio.com:5060;transport=udp>
Call-ID: 6a71eeb6-262c-123e-7fa1-7c1e523f4405
CSeq: 91735849 INVITE
Contact: <sip:gw+twilio_gateway@{LibreSBC Public IP}:5060;transport=udp;gw=twilio_gateway>
User-Agent: SBC
Allow: INVITE, ACK, BYE, CANCEL, OPTIONS, MESSAGE, INFO, UPDATE, PRACK, NOTIFY
Supported: 100rel, timer, path, replaces
Allow-Events: talk, hold, conference, refer
Privacy: none
Content-Type: application/sdp
Content-Disposition: session
Content-Length: 257
P-Asserted-Identity: "{a-number}" <sip:{a-number}@{LibreSBC Public IP}>
```

**Comment by hnimminh** [comment]
I'm really get what you trying to state, can you attach pcap or clearer explanation

**Comment by JyrgCodes** [comment]
Hi @hnimminh 
Sorry, I'll clarify. My suspicion is that LibreSBC doesn't forward the Diversion header. The request, which leaves FreePBX has the following diversion header in the request:
`Diversion: <sip:{b-number}@{FreePBX Private IP}:5060>;reason=unconditional;screen=no;privacy=off`.
And in the request, which arrives in Twilio, the Diversion header is missing

**Reply by hnimminh** [reply]
It is clear, the problem is libresbc does not propagate `Diversion`.  It is default behavior of SBC.
You need to add manipulation rule to propagate `Diversion` header

**Reply by JyrgCodes** [reply]
Hi @hnimminh 
Thanks. Could you show what that manipulation rule should look like?

**Reply by hnimminh** [reply]
something like 
```json
{
    "name": "diversion",
    "desc": "propagae diversion header",
    "actions": [
        {
            "action": "set",
            "targetvar": "sip_h_Diversion",
            "values": [
                "sip_h_Diversion"
            ]
        }
    ]
}
```

**Reply by JyrgCodes** [reply]
Hi @hnimminh 
Yeah, thanks that worked but the IP in the header is still the private one from the PBX. Could you show how to replace the IP with the public LibreSBC IP? Also, how would you conditionally run the manipulation only when the diversion header is present?

**Reply by hnimminh** [reply]
* you can do by multiple ways: 1, use `patten` & `refervar` or just manipulate iwht `values`
* same as condition
